src/tools/clippy/clippy_lints/src/dereference.rs

```diff
@@ -796,16 +796,19 @@ fn walk_parents<'tcx>(
                             },
                         })
                     }),
-                ExprKind::MethodCall(_, args, _) => {
+                ExprKind::MethodCall(_, receiver, args, _) => {
                     let id = cx.typeck_results().type_dependent_def_id(parent.hir_id).unwrap();
-                    args.iter().position(|arg| arg.hir_id == child_id).map(|i| {
-                        if i == 0 {
-                            // Check for calls to trait methods where the trait is implemented on a reference.
-                            // Two cases need to be handled:
-                            // * `self` methods on `&T` will never have auto-borrow
-                            // * `&self` methods on `&T` can have auto-borrow, but `&self` methods on `T` will take
-                            //   priority.
-                            if e.hir_id != child_id {
+                    std::iter::once(receiver)
+                        .chain(args.iter())
+                        .position(|arg| arg.hir_id == child_id)
+                        .map(|i| {
+                            if i == 0 {
```

Comment:
This branch is `receiver.hir_id == child_id`. Could we avoid the iterator combinatorics?